packages/iTwinUI-react/src/core/ComboBox/ComboBoxMenu.tsx

```diff
@@ -1,38 +1,121 @@
 import { Menu, MenuProps } from '../Menu';
-import { useSafeContext, useMergedRefs } from '../utils';
+import { Surface } from '../Surface';
+import {
+  useSafeContext,
+  useMergedRefs,
+  useVirtualization,
+  VirtualScrollProps,
+  mergeRefs,
+  getWindow,
+} from '../utils';
 import { ComboBoxStateContext, ComboBoxRefsContext } from './helpers';
 
 type ComboBoxMenuProps = Omit<MenuProps, 'onClick'> &
   React.ComponentPropsWithoutRef<'ul'>;
 
+const VirtualComboboxMenu = ({ children, ...props }: VirtualScrollProps) => {
+  const { outerProps, innerProps, visibleChildren } = useVirtualization(props);
+
+  return (
+    <div {...outerProps}>
+      {React.cloneElement(
+        children as JSX.Element,
+        {
+          ref: mergeRefs(
+            innerProps.ref,
+            (children as React.FunctionComponentElement<HTMLElement>).ref,
+          ),
+          style: {
+            ...innerProps.style,
+            ...(children as JSX.Element).props.style,
+          },
+        },
+        visibleChildren,
+      )}
+    </div>
+  );
+};
+
 export const ComboBoxMenu = React.forwardRef(
   (props: ComboBoxMenuProps, forwardedRef: React.Ref<HTMLUListElement>) => {
     const { className, style, ...rest } = props;
-    const { minWidth, id } = useSafeContext(ComboBoxStateContext);
+    const {
+      minWidth,
+      id,
+      enableVirtualization,
+      filteredOptions,
+      getMenuItem,
+      focusedIndex,
+      emptyStateContent,
+    } = useSafeContext(ComboBoxStateContext);
     const { menuRef } = useSafeContext(ComboBoxRefsContext);
 
     const refs = useMergedRefs(menuRef, forwardedRef);
 
+    const styles = React.useMemo(
+      () => ({
+        minWidth,
+        maxWidth: `min(${minWidth * 2}px, 90vw)`,
+        maxHeight: 315,
+      }),
+      [minWidth],
+    );
+
+    const overflowY = getWindow()?.CSS?.supports?.('overflow-x: overlay')
+      ? { overflowY: 'overlay' }
+      : { overflowY: 'auto' };
+
+    // 'Fool' VirtualScroll by passing length 1
+    // whenever there is no elements, to show empty state message
+    const virtualItemRenderer = React.useCallback(
+      (index: number) =>
+        filteredOptions.length > 0
+          ? getMenuItem(filteredOptions[index])
+          : emptyStateContent,
+      [emptyStateContent, filteredOptions, getMenuItem],
+    );
+
     return (
-      <Menu
-        id={`${id}-list`}
-        style={React.useMemo(
-          () => ({
-            minWidth,
-            maxWidth: `min(${minWidth * 2}px, 90vw)`,
-            maxHeight: 300,
-            ...style,
-          }),
-          [minWidth, style],
+      <>
+        {!enableVirtualization ? (
+          <Menu
+            id={`${id}-list`}
+            style={{ ...styles, ...style }}
+            setFocus={false}
+            role='listbox'
+            ref={refs}
+            className={cx('iui-scroll', className)}
+            {...rest}
+          />
+        ) : (
+          <Surface
```

Comment:
Maybe just move all virtual scroll related stuff to `VirtualComboboxMenu`

Reply:
What exactly you want me to do here? Move `Surface` and `Menu` to `VirtualizedComboboxMenu`? But then I would need to pass all the props through in there (for example `style` prop needs to be passed different for `Menu` and for `Surface`)

Reply:
Ok, I have moved everything into `VirtualizedComboBoxMenu`.
2 thins I dont like
- forwarded ref goes onto `Menu`, whereas I think it should go on `Surface`.. But different types, so not sure if I wanna hack it i.e. `mergeRefs`)
- Some repeatance of code (i.e. styles)